core/drivers/crypto/se050/core/rsa.c

```diff
@@ -0,0 +1,705 @@
+// SPDX-License-Identifier: BSD-3-Clause
+/*
+ * Copyright (C) Foundries Ltd. 2020 - All Rights Reserved
+ * Author: Jorge Ramirez <jorge@foundries.io>
+ */
+
+#include <assert.h>
+#include <drvcrypt.h>
+#include <drvcrypt_acipher.h>
+#include <drvcrypt_math.h>
+#include <initcall.h>
+#include <kernel/panic.h>
+#include <se050.h>
+#include <string.h>
+#include <tee/cache.h>
+#include <tee/tee_cryp_utl.h>
+
+static uint32_t tee2se050(uint32_t algo)
+{
+	switch (algo) {
+	case TEE_ALG_RSASSA_PKCS1_V1_5_SHA1:
+		return kAlgorithm_SSS_RSASSA_PKCS1_V1_5_SHA1;
+	case TEE_ALG_RSASSA_PKCS1_V1_5_SHA224:
+		return kAlgorithm_SSS_RSASSA_PKCS1_V1_5_SHA224;
+	case TEE_ALG_RSASSA_PKCS1_V1_5_SHA256:
+		return kAlgorithm_SSS_RSASSA_PKCS1_V1_5_SHA256;
+	case TEE_ALG_RSASSA_PKCS1_V1_5_SHA384:
+		return kAlgorithm_SSS_RSASSA_PKCS1_V1_5_SHA384;
+	case TEE_ALG_RSASSA_PKCS1_V1_5_SHA512:
+		return kAlgorithm_SSS_RSASSA_PKCS1_V1_5_SHA512;
+	case TEE_ALG_RSASSA_PKCS1_PSS_MGF1_SHA1:
+		return kAlgorithm_SSS_RSASSA_PKCS1_PSS_MGF1_SHA1;
+	case TEE_ALG_RSASSA_PKCS1_PSS_MGF1_SHA224:
+		return kAlgorithm_SSS_RSASSA_PKCS1_PSS_MGF1_SHA224;
+	case TEE_ALG_RSASSA_PKCS1_PSS_MGF1_SHA256:
+		return kAlgorithm_SSS_RSASSA_PKCS1_PSS_MGF1_SHA256;
+	case TEE_ALG_RSASSA_PKCS1_PSS_MGF1_SHA384:
+		return kAlgorithm_SSS_RSASSA_PKCS1_PSS_MGF1_SHA384;
+	case TEE_ALG_RSASSA_PKCS1_PSS_MGF1_SHA512:
+		return kAlgorithm_SSS_RSASSA_PKCS1_PSS_MGF1_SHA512;
+	case TEE_ALG_RSAES_PKCS1_V1_5:
+		return kAlgorithm_SSS_RSAES_PKCS1_V1_5_SHA256;
+	case TEE_ALG_RSAES_PKCS1_OAEP_MGF1_SHA1:
+		return kAlgorithm_SSS_RSAES_PKCS1_OAEP_SHA1;
+	case TEE_ALG_RSAES_PKCS1_OAEP_MGF1_SHA224:
+		return kAlgorithm_SSS_RSAES_PKCS1_OAEP_SHA224;
+	case TEE_ALG_RSAES_PKCS1_OAEP_MGF1_SHA256:
+		return kAlgorithm_SSS_RSAES_PKCS1_OAEP_SHA256;
+	case TEE_ALG_RSAES_PKCS1_OAEP_MGF1_SHA384:
+		return kAlgorithm_SSS_RSAES_PKCS1_OAEP_SHA384;
+	case TEE_ALG_RSAES_PKCS1_OAEP_MGF1_SHA512:
+		return kAlgorithm_SSS_RSAES_PKCS1_OAEP_SHA512;
+	case TEE_ALG_RSA_NOPAD:
+		return kAlgorithm_SSS_RSASSA_NO_PADDING;
+#ifdef CFG_CRYPTO_RSASSA_NA1
+	case TEE_ALG_RSASSA_PKCS1_V1_5:
+#endif
+	case TEE_ALG_RSASSA_PKCS1_V1_5_MD5:
+	case TEE_ALG_RSASSA_PKCS1_V1_5_MD5SHA1:
+	default:
+		EMSG("rsa algorithm 0x%x not supported ", algo);
+		panic();
+		return kAlgorithm_None;
+	}
+}
+
+static bool bn_alloc_max(struct bignum **s)
+{
+	*s = crypto_bignum_allocate(4096);
+
+	return *s;
```

Comment:
`return crypto_bignum_allocate(4096);`

Reply:
this just borrows the implementation from ltc helpers. returning what you propose will not set the value in the pointer.

Reply:
right. discard my comment.

Reply:
It's not a good idea to do the allocation with the maximum 4096 bits if you don't need all this memory for each field of the key. Allocation can be optimized.

Reply:
just doing the same than libtomcrypt/rsa.c. could you send a patch for that implementation (tomcrypt) and I''ll then replace mine please? I dont quite know what you have in mind...

Reply:
Have a look to this NIST specification to understand RSA key fields and size: https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-56Br2.pdf
Know memory is not extendable in embedded system so you may understand that allocating unused memory is something that might generate memory allocation error whereas part of the memory is not used. Why do you think the key size is one of the function parameters if it's to not used it?
That say, I have no patch for tomcrypt but I have done the key memory allocation in the core/driver/crypto/caam/acipher/caam_rsa.c. I would not say this is the best but this one is more optimized than tomcrypt ones.